src/adapter/src/repositories/in_memory/cache.rs

```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    /// Removes expired entries from the cache asynchronously.
+    ///
+    /// This method is used to remove any expired cache entries based on their expiration time.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the cleanup is successful, otherwise returns a `CoreError`.
+    async fn cleanup_expired_entries(&self) -> Result<(), CoreError> {
+        let mut cache = self.cache.write().await;
+        let now = SystemTime::now();
+        cache.retain(|_, (_, expiry_time, _)| *expiry_time > now);
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl CachePort for InMemoryCache {
+    /// Retrieves a value from the cache based on the provided key.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: A string representing the key to retrieve from the cache.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` containing `Some(value)` if the key is found and not expired,
+    /// or `None` if the key is not found or expired. Returns a `CoreError` on failure.
+    async fn get(&self, key: &str) -> Result<Option<String>, CoreError> {
+        let cache = self.cache.read().await;
+        Ok(cache
+            .get(key)
+            .filter(|(_, expiry_time, _)| *expiry_time > SystemTime::now())
+            .map(|(value, _, _)| value.clone()))
+    }
+
+    /// Sets a key-value pair in the cache with an optional expiration duration.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: A string representing the key to set in the cache.
+    /// * `value`: A string representing the value to associate with the key.
+    /// * `expiration`: An optional `Duration` specifying the expiration time for the key-value pair.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the set operation is successful, otherwise returns a `CoreError`.
+    async fn set(
+        &mut self,
+        key: String,
+        value: String,
+        expiration: Option<Duration>,
+    ) -> Result<bool, CoreError> {
+        self.cleanup_expired_entries().await?;
+        let expiry_time = expiration.map(|exp| SystemTime::now() + exp);
+        let mut cache = self.cache.write().await;
+        cache.insert(
+            key,
+            (
+                value,
+                expiry_time.unwrap_or_else(|| SystemTime::UNIX_EPOCH),
+                expiration,
+            ),
+        );
+        Ok(true)
+    }
+
+    /// Removes a key-value pair from the cache based on the provided key.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: A string representing the key to remove from the cache.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the key is found and removed, otherwise returns a `CoreError`.
+    async fn remove(&mut self, key: &str) -> Result<bool, CoreError> {
+        let mut cache = self.cache.write().await;
+        if cache.remove(key).is_some() {
+            Ok(true)
+        } else {
+            Err(CoreError::NotFound)
```

Comment:
Not match the behavior of Redis `Ok(false)`.

Reply:
I want the `NotFound` error has same meaning to `false` response.